fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -1858,22 +1858,62 @@ struct RocksDBKeyValueStore : IKeyValueStore {
 	void set(KeyValueRef kv, const Arena*) override {
 		if (writeBatch == nullptr) {
 			writeBatch.reset(new rocksdb::WriteBatch());
+			setKeysSet.clear();
 		}
 		ASSERT(defaultFdbCF != nullptr);
 		writeBatch->Put(defaultFdbCF, toSlice(kv.key), toSlice(kv.value));
+		if (SERVER_KNOBS->ROCKSDB_SINGLEKEY_DELETES_ON_CLEARRANGE) {
+			setKeysSet.insert(kv.key);
+		}
+		TraceEvent("DBEventSet").detail("Key", kv.key.toString());
 	}
 
-	void clear(KeyRangeRef keyRange, const Arena*) override {
+	void clear(KeyRangeRef keyRange, const StorageServerMetrics* storageMetrics, const Arena*) override {
 		if (writeBatch == nullptr) {
 			writeBatch.reset(new rocksdb::WriteBatch());
+			setKeysSet.clear();
 		}
+		TraceEvent("DBEventClear").detail("KeyBegin", keyRange.begin).detail("KeyEnd", keyRange.end);
 
 		ASSERT(defaultFdbCF != nullptr);
 
 		if (keyRange.singleKeyRange()) {
 			writeBatch->Delete(defaultFdbCF, toSlice(keyRange.begin));
 		} else {
-			writeBatch->DeleteRange(defaultFdbCF, toSlice(keyRange.begin), toSlice(keyRange.end));
+			if (SERVER_KNOBS->ROCKSDB_SINGLEKEY_DELETES_ON_CLEARRANGE && storageMetrics != nullptr &&
+			    storageMetrics->byteSample.getEstimate(keyRange) <
+			        SERVER_KNOBS->ROCKSDB_SINGLEKEY_DELETES_BYTES_LIMIT) {
+				rocksdb::ReadOptions options;
+				options.background_purge_on_iterator_cleanup = true;
+				options.auto_prefix_mode = (SERVER_KNOBS->ROCKSDB_PREFIX_LEN > 0);
+				auto beginSlice = toSlice(keyRange.begin);
+				auto endSlice = toSlice(keyRange.end);
+				options.iterate_lower_bound = &beginSlice;
+				options.iterate_upper_bound = &endSlice;
+				auto cursor = std::unique_ptr<rocksdb::Iterator>(db->NewIterator(options, defaultFdbCF));
+				cursor->Seek(toSlice(keyRange.begin));
+				int numKeys = 0;
+				while (cursor->Valid() && toStringRef(cursor->key()) < keyRange.end) {
+					writeBatch->Delete(defaultFdbCF, cursor->key());
+					cursor->Next();
+					numKeys++;
+				}
+
+				int numV = 0;
+				for (auto key : setKeysSet) {
+					TraceEvent("SetKeySets").detail("Key", key);
+					if (StringRef(key) >= keyRange.begin && StringRef(key) < keyRange.end) {
+						writeBatch->Delete(defaultFdbCF, toSlice(key));
+						numV++;
+					} else {
+						break;
```

Comment:
I think this break could be too early: suppose the keyset has: {1, 2, 3, 4, 5}, and the keyRange is [3, 5).

```
while (StringRef(key) < keyRange.end) {
  if (StringRef(key) >= keyRange.begin) { // Delete it
```

In addition, std::set is an ordered set, you should be able to use `std::lower_bound` to find the first key in the range of `keyRange`